File: frontend/src/optical/kinds/aom-v3/physics.ts
/**
 * AOM (Acousto-Optic Modulator) PhysicsOp — v3.
 *
 * Op name: `diffract_aom`
 * Kind:    `aom`
 *
 * Bidirectional diffraction faces:
 *   - Canonical v3 two-port layout is physical faces A/B.
 *   - Directed transitions A -> B and B -> A encode opposite
 *     frequency/side branches unless transition.params.order explicitly
 *     overrides them.
 *   - Older A1/B1/A2/B2 catalog rows are still accepted.
 *   - Legacy B0/B+1/B-1 out faces are still accepted for old catalog rows.
 *
 * Geometry (asset body convention):
 *   - Body +z = optical axis (input direction)
 *   - Body +x = acoustic propagation direction (RF traveling wave)
 *   - rfPropagationDirectionBodyLocal is a vector, not an optical face.
 *   - RF/acoustic vector must be perpendicular to the transition A->B axis.
 *
 * Bragg angle:
 *   Current convention: theta_B = asin(lambda * f / (2 * v_acoustic)).
 *   This is the external half-angle; n is used for slab/q propagation.
 *   theta_B = asin(lambda * f / (2 * v_acoustic))
 *
 * Order efficiency (simple model; refine in Phase 6 with a Bragg efficiency curve):
 *   - +1:    eta = baseEfficiency
 *   -  0:    1 - eta  (undiffracted leftover)
 *   - -1:    eta * suppressedFraction  (wrong-sign diffraction, usually < 1%)
 *   - other: 0 (clip)
 *
 * Frequency source priority:
 *   Dynamic RF values should be resolved from the actual RF object chain.
 *   RF amplifier gain belongs upstream; AOM consumes post-chain RF vector signal.
 *   ctx.dynamic.aomFreqMhz -> ctx.params.aomFreqMhz -> ctx.params.centerFreqMhz
 *
 * Polarization & q-parameter:
 *   - Jones unchanged (simple model: AOMs are weakly birefringent)
 *   - q propagates through crystal slab: B = L / n
 */

import {
  type BeamRay,
  type Vec3,
  vec3Distance,
} from "../../beam-ray";
import { type Complex, cAdd, cDiv } from "../../fiber/gaussian";
import {
  type PhysicsOp,
  type PhysicsOpContext,
  registerKind,
} from "../../registry";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function applyAbcdToQ(A: number, B: number, C: number, D: number, q: Complex): Complex {
  const Aq: Complex = { re: A * q.re, im: A * q.im };
  const Bc: Complex = { re: B, im: 0 };
  const Cq: Complex = { re: C * q.re, im: C * q.im };
  const Dc: Complex = { re: D, im: 0 };
  return cDiv(cAdd(Aq, Bc), cAdd(Cq, Dc));
}

const RF_LOAD_Z_OHM = 50.0;

const clamp01 = (x: number): number => Math.max(0, Math.min(1, x));

function finiteNumber(v: unknown): number | undefined {
  return typeof v === "number" && Number.isFinite(v) ? v : undefined;
}

function positiveFiniteNumber(v: unknown): number | undefined {
  const n = finiteNumber(v);
  return n !== undefined && n > 0 ? n : undefined;
}

function vppToPowerW(vpp: number, zOhm: number = RF_LOAD_Z_OHM): number {
  return (vpp * vpp) / (8 * zOhm);
}

function dbmToW(dbm: number): number {
  return Math.pow(10, (dbm - 30) / 10);
}

function readRfFrequencyMhz(ctx: PhysicsOpContext): number {
  return (
    positiveFiniteNumber(ctx.dynamic?.aomFreqMhz)
    ?? positiveFiniteNumber(ctx.dynamic?.rfFrequencyMhz)
    ?? positiveFiniteNumber(ctx.dynamic?.aomRfFreqMhz)
    ?? positiveFiniteNumber(ctx.params.aomFreqMhz)
    ?? positiveFiniteNumber(ctx.params.centerFreqMhz)
    ?? 80
  );
}

function readRfDrivePowerW(ctx: PhysicsOpContext): number | undefined {
  let p =
    finiteNumber(ctx.dynamic?.rfDrivePowerW)
    ?? finiteNumber(ctx.dynamic?.aomRfPowerW)
    ?? finiteNumber(ctx.params.rfDrivePowerW)
    ?? finiteNumber(ctx.params.aomRfPowerW);

  const vpp =
    positiveFiniteNumber(ctx.dynamic?.aomRfVpp)
    ?? positiveFiniteNumber(ctx.dynamic?.rfVpp)
    ?? positiveFiniteNumber(ctx.params.aomRfVpp)
    ?? positiveFiniteNumber(ctx.params.rfVpp);
  if (p === undefined && vpp !== undefined) {
    p = vppToPowerW(vpp);
  }

  const dbm =
    finiteNumber(ctx.dynamic?.aomRfPowerDbm)
    ?? finiteNumber(ctx.dynamic?.rfPowerDbm)
    ?? finiteNumber(ctx.params.aomRfPowerDbm)
    ?? finiteNumber(ctx.params.rfPowerDbm);
  if (p === undefined && dbm !== undefined) {
    p = dbmToW(dbm);
  }

  if (p === undefined || !Number.isFinite(p) || p < 0) return undefined;
  const maxW = positiveFiniteNumber(ctx.params.rfPowerMaxW);
  return maxW !== undefined ? Math.min(p, maxW) : p;
}

function vecDot(a: Vec3, b: Vec3): number {
  return a.x * b.x + a.y * b.y + a.z * b.z;
}

function vecLength(a: Vec3): number {
  return Math.sqrt(vecDot(a, a));
}

function vecScale(a: Vec3, s: number): Vec3 {
  return { x: a.x * s, y: a.y * s, z: a.z * s };
}

function vecAdd(a: Vec3, b: Vec3): Vec3 {
  return { x: a.x + b.x, y: a.y + b.y, z: a.z + b.z };
}

function vecSub(a: Vec3, b: Vec3): Vec3 {
  return { x: a.x - b.x, y: a.y - b.y, z: a.z - b.z };
}

function vecNormalize(a: Vec3): Vec3 {
  const len = vecLength(a);
  if (len < 1e-15) return { x: 1, y: 0, z: 0 };
  return vecScale(a, 1 / len);
}

function readVec3(v: unknown): Vec3 | undefined {
  let x: unknown;
  let y: unknown;
  let z: unknown;
  if (Array.isArray(v) && v.length >= 3) {
    [x, y, z] = v;
  } else if (v && typeof v === "object") {
    ({ x, y, z } = v as { x?: unknown; y?: unknown; z?: unknown });
  } else {
    return undefined;
  }
  if (
    typeof x !== "number" || !Number.isFinite(x)
    || typeof y !== "number" || !Number.isFinite(y)
    || typeof z !== "number" || !Number.isFinite(z)
  ) {
    return undefined;
  }
  return { x, y, z };
}

function transitionOpticalAxis(ctx: PhysicsOpContext): Vec3 {
  return vecNormalize(
    ctx.faceOut.normalBodyLocal
    ?? vecSub(ctx.faceOut.positionMmBodyLocal, ctx.faceIn.positionMmBodyLocal),
  );
}

function readRfDirectionBodyLocal(ctx: PhysicsOpContext): Vec3 {
  const rawRfDir =
    readVec3(ctx.dynamic?.rfPropagationDirectionBodyLocal)
    ?? readVec3(ctx.dynamic?.acousticAxisBodyLocal)
    ?? readVec3(ctx.params.rfPropagationDirectionBodyLocal)
    ?? readVec3(ctx.params.acousticAxisBodyLocal)
    ?? { x: 1, y: 0, z: 0 };
  const rfLen = vecLength(rawRfDir);
  if (rfLen < 1e-15) {
    throw new Error("AOM rfPropagationDirectionBodyLocal must be a non-zero vector");
  }
  const rfDir = vecScale(rawRfDir, 1 / rfLen);
  const opticalAxis = transitionOpticalAxis(ctx);
  const dot = vecDot(rfDir, opticalAxis);
  if (Math.abs(dot) > 1e-6) {
    throw new Error(
      `AOM rfPropagationDirectionBodyLocal must be perpendicular to the A->B optical axis; dot=${dot}`,
    );
  }
  return rfDir;
}

/** Parse legacy order m from a face id matching `B<signed_int>`. */
export function parseOrderFromFaceId(faceId: string): number {
  const m = /^B([+-]?\d+)$/.exec(faceId);
  if (!m) {
    throw new Error(
      `AOM diffract_aom: faceOut.id must match B[+/-n], got "${faceId}"`,
    );
  }
  return parseInt(m[1]!, 10);
}

export function orderFromContext(ctx: PhysicsOpContext): number {
  const explicit =
    finiteNumber(ctx.params.order)
    ?? finiteNumber(ctx.params.diffractionOrder);
  if (explicit !== undefined) return Math.trunc(explicit);

  if (ctx.faceIn.id === "A" && ctx.faceOut.id === "B") return 1;
  if (ctx.faceIn.id === "B" && ctx.faceOut.id === "A") return -1;
  if (ctx.faceIn.id === "A1" && ctx.faceOut.id === "B1") return 1;
  if (ctx.faceIn.id === "A2" && ctx.faceOut.id === "B2") return -1;
  return parseOrderFromFaceId(ctx.faceOut.id);
}

/** Bragg angle in radians.
 *  theta_B = asin(lambda * f / (2 * v)) with lambda in nm -> m and f in MHz -> Hz. */
export function braggAngleRad(
  wavelengthNm: number,
  freqMhz: number,
  acousticVelocityMps: number,
  _refractiveIndex?: number,
): number {
  const lambdaM = wavelengthNm * 1e-9;
  const fHz = freqMhz * 1e6;
  const arg = (lambdaM * fHz) / (2 * acousticVelocityMps);
  // Clamp to [-1, 1] to avoid asin overflow when params push beyond Bragg.
  return Math.asin(Math.max(-1, Math.min(1, arg)));
}

function deflectAlongRfSide(inputDir: Vec3, rfDir: Vec3, deflectRad: number): Vec3 {
  const inDir = vecNormalize(inputDir);
  const rfTransverse = vecSub(rfDir, vecScale(inDir, vecDot(rfDir, inDir)));
  const side = vecNormalize(rfTransverse);
  return vecNormalize(vecAdd(
    vecScale(inDir, Math.cos(deflectRad)),
    vecScale(side, Math.sin(deflectRad)),
  ));
}

/** Order efficiency split. Refined Phase 6+ with a Bragg efficiency curve. */
export function orderEfficiency(order: number, baseEfficiency: number): number {
  const eta = clamp01(baseEfficiency);
  if (order === 1) return eta;
  if (order === 0) return Math.max(0, 1 - eta);
  if (order === -1) return eta * 0.01;       // ~1% wrong-sign suppression
  return 0;
}

export function firstOrderEfficiencyFromContext(
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
  thetaBRad: number,
): number {
  const rfPowerW = readRfDrivePowerW(ctx);
  const requiresRfDrive = ctx.params.requiresRfDrive === true;
  if (requiresRfDrive && rfPowerW === undefined) return 0;

  const m2 = positiveFiniteNumber(ctx.params.figureOfMeritM2);
  const Lmm = positiveFiniteNumber(ctx.params.crystalLengthMm);
  const Wmm = positiveFiniteNumber(ctx.params.acousticBeamWidthMm);
  if (rfPowerW !== undefined && m2 !== undefined && Lmm !== undefined && Wmm !== undefined) {
    const lambdaM = rayIn.wavelengthNm * 1e-9;
    const Lm = Lmm * 1e-3;
    const Wm = Wmm * 1e-3;
    const inner = Math.sqrt((2 * m2 * rfPowerW) / Wm);
    const arg = ((Math.PI * Lm) / (2 * lambdaM * Math.cos(thetaBRad))) * inner;
    return clamp01(Math.sin(arg) ** 2);
  }

  return clamp01(finiteNumber(ctx.params.baseEfficiency) ?? 0.85);
}

/** Unsigned angle between a ray direction and the optical axis. */
function angleToAxisRad(direction: Vec3, axis: Vec3): number {
  const d = vecNormalize(direction);
  const a = vecNormalize(axis);
  return Math.acos(Math.max(-1, Math.min(1, vecDot(d, a))));
}

/**
 * sinc^2 Bragg phase-matching factor for off-axis incidence.
 *
 * Thick-grating coupled-wave theory: efficiency scales as sinc^2(dk·L/2),
 * where dk is the longitudinal wave-vector mismatch produced by deviating
 * from the matched incidence angle. The canonical on-axis input (ray along
 * the A->B optical axis) is treated as perfectly matched, so a ray arriving
 * at external angle dthetaExt sees:
 *   K       = 2π·f / v               (acoustic grating vector, 1/m)
 *   thetaB  = asin(λ·f / (2·n·v))    (internal Bragg angle)
 *   dk      = K·cos(thetaB)·(dthetaExt / n)
 *   xi      = dk·L / 2
 *   factor  = (sin xi / xi)^2        (→ 1 as xi → 0)
 */
export function braggDetuningFactor(
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
  freqMhz: number,
  vAcoustic: number,
  n: number,
  Lmm: number,
): number {
  const dthetaExt = angleToAxisRad(rayIn.direction, transitionOpticalAxis(ctx));
  if (dthetaExt === 0) return 1;
  const lambdaM = rayIn.wavelengthNm * 1e-9;
  const fHz = freqMhz * 1e6;
  const Lm = Lmm * 1e-3;
  const kAcoustic = (2 * Math.PI * fHz) / vAcoustic;
  const thetaBInt = Math.asin(
    Math.max(-1, Math.min(1, (lambdaM * fHz) / (2 * n * vAcoustic))),
  );
  const dthetaInt = dthetaExt / n;
  const dk = kAcoustic * Math.cos(thetaBInt) * dthetaInt;
  const xi = (dk * Lm) / 2;
  if (xi === 0) return 1;
  return clamp01((Math.sin(xi) / xi) ** 2);
}

/** External half-width to the first sinc^2 null (xi = π), in mrad:
 *  dthetaExt = n·v / (f·L). */
export function braggAcceptanceMrad(
  freqMhz: number,
  vAcoustic: number,
  n: number,
  Lmm: number,
): number {
  const fHz = freqMhz * 1e6;
  const Lm = Lmm * 1e-3;
  return ((n * vAcoustic) / (fHz * Lm)) * 1e3;
}

// ---------------------------------------------------------------------------
// diffract_aom op
// ---------------------------------------------------------------------------

export const diffractAomOp: PhysicsOp = (
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
): BeamRay[] => {
  const order = orderFromContext(ctx);

  const freqMhz = readRfFrequencyMhz(ctx);
  const vAcoustic =
    positiveFiniteNumber(ctx.params.acousticVelocityMps)
    ?? positiveFiniteNumber(ctx.params.acousticVelocityMPerS)
    ?? 4200;
  const n = positiveFiniteNumber(ctx.params.refractiveIndex) ?? 2.26;
  const L = positiveFiniteNumber(ctx.params.crystalLengthMm)
    ?? vec3Distance(ctx.faceIn.positionMmBodyLocal, ctx.faceOut.positionMmBodyLocal);

  const theta_B = braggAngleRad(rayIn.wavelengthNm, freqMhz, vAcoustic, n);
  const deflectRad = order * 2 * theta_B;

  const dirOut = deflectAlongRfSide(
    rayIn.direction,
    readRfDirectionBodyLocal(ctx),
    deflectRad,
  );

  const detune = braggDetuningFactor(rayIn, ctx, freqMhz, vAcoustic, n, L);
  const firstOrderEfficiency = firstOrderEfficiencyFromContext(rayIn, ctx, theta_B) * detune;
  const eff = orderEfficiency(order, firstOrderEfficiency);
  const newPower = rayIn.powerMw * eff;

  // Doppler shift: order m diffracts off the f_RF acoustic wave, shifting the
  // optical frequency by m·f_RF. Tracked as an offset on the nominal carrier.
  const newFreqOffsetHz = (rayIn.freqOffsetHz ?? 0) + order * freqMhz * 1e6;

  // q-parameter slab propagation (B = L/n inside the crystal).
  const Bslab = L / n;
  const qxOut = applyAbcdToQ(1, Bslab, 0, 1, rayIn.qx);
  const qyOut = applyAbcdToQ(1, Bslab, 0, 1, rayIn.qy);

  // Chief ray exits at faceOut position (or its projection along the
  // deflected direction).
  const newOrigin = ctx.faceOut.positionMmBodyLocal;
  const thickness = vec3Distance(
    ctx.faceIn.positionMmBodyLocal, ctx.faceOut.positionMmBodyLocal,
  );

  return [{
    ...rayIn,
    origin: newOrigin,
    direction: dirOut,
    qx: qxOut,
    qy: qyOut,
    powerMw: newPower,
    pathLengthMm: rayIn.pathLengthMm + thickness,
    freqOffsetHz: newFreqOffsetHz,
    // jones unchanged (simple polarization-preserving model)
  }];
};

registerKind("aom", {
  ops: {
    diffract_aom: diffractAomOp,
  },
  needsAperture: true,
});

File: backend/app/optical/kinds/aom/physics.py
"""AOM v3 PhysicsOp (Python mirror of frontend kinds/aom-v3/physics.ts).

Op name: diffract_aom
Kind: aom
Canonical v3 two-port layout is physical faces A/B. Directed transitions
A -> B and B -> A encode opposite frequency/side branches unless
transition params explicitly set order. Older A1/B1/A2/B2 catalog rows
and legacy B0/B+1/B-1 out faces are still accepted.
"""

from __future__ import annotations

import math
import re
from typing import Optional

from app.optical.aom_physics import (
    acoustic_incidence_rad,
    bragg_order_detune,
    first_order_efficiency,
    order_efficiency,
)
from app.optical.beam_ray import BeamRay, Vec3, vec3_distance
from app.optical.registry import (
    KindEntry,
    PhysicsOpContext,
    register_kind,
)


def _apply_abcd_to_q(A: float, B: float, C: float, D: float, q: complex) -> complex:
    return (A * q + B) / (C * q + D)


_FACE_ID_PATTERN = re.compile(r"^B([+-]?\d+)$")
RF_LOAD_Z_OHM = 50.0


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, x))


def _finite_number(value) -> Optional[float]:
    if isinstance(value, (int, float)) and math.isfinite(float(value)):
        return float(value)
    return None


def _positive_finite_number(value) -> Optional[float]:
    n = _finite_number(value)
    if n is not None and n > 0:
        return n
    return None


def _vec_dot(a: Vec3, b: Vec3) -> float:
    return a.x * b.x + a.y * b.y + a.z * b.z


def _vec_len(a: Vec3) -> float:
    return math.sqrt(_vec_dot(a, a))


def _vec_tuple(a: Vec3) -> tuple[float, float, float]:
    return (a.x, a.y, a.z)


def _vec_norm(a: Vec3) -> Vec3:
    n = _vec_len(a)
    if n < 1e-15:
        return Vec3(1, 0, 0)
    return Vec3(a.x / n, a.y / n, a.z / n)


def _read_vec3(value) -> Optional[Vec3]:
    if isinstance(value, (list, tuple)) and len(value) >= 3:
        x, y, z = value[:3]
    elif isinstance(value, dict):
        x, y, z = value.get("x"), value.get("y"), value.get("z")
    else:
        return None
    if all(isinstance(v, (int, float)) and math.isfinite(float(v)) for v in (x, y, z)):
        return Vec3(float(x), float(y), float(z))
    return None


def _transition_optical_axis(ctx: PhysicsOpContext) -> Vec3:
    if ctx.face_out.normal_body_local is not None:
        return _vec_norm(ctx.face_out.normal_body_local)
    return _vec_norm(Vec3(
        ctx.face_out.position_mm_body_local.x - ctx.face_in.position_mm_body_local.x,
        ctx.face_out.position_mm_body_local.y - ctx.face_in.position_mm_body_local.y,
        ctx.face_out.position_mm_body_local.z - ctx.face_in.position_mm_body_local.z,
    ))


def _read_rf_direction_body_local(ctx: PhysicsOpContext) -> Vec3:
    dynamic = ctx.dynamic or {}
    raw_rf_dir = (
        _read_vec3(dynamic.get("rfPropagationDirectionBodyLocal"))
        or _read_vec3(dynamic.get("acousticAxisBodyLocal"))
        or _read_vec3(ctx.params.get("rfPropagationDirectionBodyLocal"))
        or _read_vec3(ctx.params.get("acousticAxisBodyLocal"))
        or Vec3(1, 0, 0)
    )
    rf_len = _vec_len(raw_rf_dir)
    if rf_len < 1e-15:
        raise ValueError("AOM rfPropagationDirectionBodyLocal must be a non-zero vector")
    rf_dir = Vec3(raw_rf_dir.x / rf_len, raw_rf_dir.y / rf_len, raw_rf_dir.z / rf_len)
    optical_axis = _transition_optical_axis(ctx)
    dot = _vec_dot(rf_dir, optical_axis)
    if abs(dot) > 1e-6:
        raise ValueError(
            "AOM rfPropagationDirectionBodyLocal must be perpendicular "
            f"to the A->B optical axis; dot={dot}"
        )
    return rf_dir


def _first_defined(*values: Optional[float]) -> Optional[float]:
    for value in values:
        if value is not None:
            return value
    return None


def _vpp_to_power_w(vpp: float, z_ohm: float = RF_LOAD_Z_OHM) -> float:
    return (vpp * vpp) / (8.0 * z_ohm)


def _dbm_to_w(dbm: float) -> float:
    return 10.0 ** ((dbm - 30.0) / 10.0)


def _read_rf_frequency_mhz(ctx: PhysicsOpContext) -> float:
    dynamic = ctx.dynamic or {}
    return (
        _positive_finite_number(dynamic.get("aomFreqMhz"))
        or _positive_finite_number(dynamic.get("rfFrequencyMhz"))
        or _positive_finite_number(dynamic.get("aomRfFreqMhz"))
        or _positive_finite_number(ctx.params.get("aomFreqMhz"))
        or _positive_finite_number(ctx.params.get("centerFreqMhz"))
        or 80.0
    )


def _read_rf_drive_power_w(ctx: PhysicsOpContext) -> Optional[float]:
    dynamic = ctx.dynamic or {}
    p = _first_defined(
        _finite_number(dynamic.get("rfDrivePowerW")),
        _finite_number(dynamic.get("aomRfPowerW")),
        _finite_number(ctx.params.get("rfDrivePowerW")),
        _finite_number(ctx.params.get("aomRfPowerW")),
    )
    vpp = _first_defined(
        _positive_finite_number(dynamic.get("aomRfVpp")),
        _positive_finite_number(dynamic.get("rfVpp")),
        _positive_finite_number(ctx.params.get("aomRfVpp")),
        _positive_finite_number(ctx.params.get("rfVpp")),
    )
    if p is None and vpp is not None:
        p = _vpp_to_power_w(vpp)
    dbm = _first_defined(
        _finite_number(dynamic.get("aomRfPowerDbm")),
        _finite_number(dynamic.get("rfPowerDbm")),
        _finite_number(ctx.params.get("aomRfPowerDbm")),
        _finite_number(ctx.params.get("rfPowerDbm")),
    )
    if p is None and dbm is not None:
        p = _dbm_to_w(dbm)
    if p is None or not math.isfinite(p) or p < 0:
        return None
    max_w = _positive_finite_number(ctx.params.get("rfPowerMaxW"))
    return min(p, max_w) if max_w is not None else p


def parse_order_from_face_id(face_id: str) -> int:
    m = _FACE_ID_PATTERN.match(face_id)
    if not m:
        raise ValueError(
            f'AOM diffract_aom: face_out.id must match B[+/-n], got "{face_id}"'
        )
    return int(m.group(1))


def order_from_context(ctx: PhysicsOpContext) -> int:
    explicit = _first_defined(
        _finite_number(ctx.params.get("order")),
        _finite_number(ctx.params.get("diffractionOrder")),
    )
    if explicit is not None:
        return int(explicit)
    if ctx.face_in.id == "A" and ctx.face_out.id == "B":
        return 1
    if ctx.face_in.id == "B" and ctx.face_out.id == "A":
        return -1
    if ctx.face_in.id == "A1" and ctx.face_out.id == "B1":
        return 1
    if ctx.face_in.id == "A2" and ctx.face_out.id == "B2":
        return -1
    return parse_order_from_face_id(ctx.face_out.id)


def bragg_angle_rad(
    wavelength_nm: float,
    freq_mhz: float,
    acoustic_velocity_mps: float,
    refractive_index: float | None = None,
) -> float:
    """asin(lambda * f / (2 * v)) with lambda in nm -> m and f in MHz -> Hz."""
    lambda_m = wavelength_nm * 1e-9
    f_hz = freq_mhz * 1e6
    _ = refractive_index
    arg = (lambda_m * f_hz) / (2 * acoustic_velocity_mps)
    return math.asin(max(-1.0, min(1.0, arg)))


def _deflect_along_rf_side(input_dir: Vec3, rf_dir: Vec3, deflect_rad: float) -> Vec3:
    in_dir = _vec_norm(input_dir)
    dot = _vec_dot(rf_dir, in_dir)
    rf_transverse = Vec3(
        rf_dir.x - in_dir.x * dot,
        rf_dir.y - in_dir.y * dot,
        rf_dir.z - in_dir.z * dot,
    )
    side = _vec_norm(rf_transverse)
    return _vec_norm(Vec3(
        in_dir.x * math.cos(deflect_rad) + side.x * math.sin(deflect_rad),
        in_dir.y * math.cos(deflect_rad) + side.y * math.sin(deflect_rad),
        in_dir.z * math.cos(deflect_rad) + side.z * math.sin(deflect_rad),
    ))


def bragg_detuning_factor(
    ray_in: BeamRay,
    ctx: PhysicsOpContext,
    order: int,
    theta_b_rad: float,
    freq_mhz: float,
    v_acoustic: float,
    n: float,
    l_mm: float,
) -> float:
    """sinc^2 Bragg phase-matching factor for THIS order at the actual tilt.

    Order m is matched when the signed incidence about the acoustic axis is
    ``-m * theta_B`` (``aom_physics.bragg_matched_incidence_rad``), so the two
    +/-1 orders peak at tilts 2*theta_B apart. Mirrors the production anchor op
    (``anchor_ops/aom.py``) — both go through ``aom_physics`` so the two trace
    paths can't drift.
    """
    theta_in = acoustic_incidence_rad(
        (ray_in.direction.x, ray_in.direction.y, ray_in.direction.z),
        _vec_tuple(_read_rf_direction_body_local(ctx)),
        _vec_tuple(_transition_optical_axis(ctx)),
    )
    if theta_in is None:
        return 1.0
    return bragg_order_detune(
        order, theta_in, theta_b_rad,
        ray_in.wavelength_nm, freq_mhz, v_acoustic, n, l_mm,
    )


def bragg_acceptance_mrad(freq_mhz: float, v_acoustic: float, n: float, l_mm: float) -> float:
    """External half-width to the first sinc^2 null (xi = pi), in mrad.

    Solving xi = dk*L/2 = pi with dk = K*(dtheta_ext/n) (cos(theta_B)~1):
        dtheta_ext = n * v / (f * L)
    """
    f_hz = freq_mhz * 1e6
    l_m = l_mm * 1e-3
    return (n * v_acoustic / (f_hz * l_m)) * 1e3


def first_order_efficiency_from_context(
    ray_in: BeamRay,
    ctx: PhysicsOpContext,
    theta_b_rad: float,
) -> float:
    # Delegate to the shared datasheet-calibrated model (single source of truth,
    # also used by the production anchor op). theta_b_rad is unused now (the
    # model is normalized on P_peak, not M2/L/W).
    _ = theta_b_rad

    def _fp(key: str, default: float) -> float:
        v = _positive_finite_number(ctx.params.get(key))
        return v if v is not None else default

    return first_order_efficiency(
        wavelength_nm=ray_in.wavelength_nm,
        freq_mhz=_read_rf_frequency_mhz(ctx),
        rf_power_w=_read_rf_drive_power_w(ctx),
        peak_efficiency=_finite_number(ctx.params.get("baseEfficiency")) or 0.85,
        rf_power_for_peak_w=_fp("rfPowerForPeakW", 2.2),
        peak_ref_wavelength_nm=_fp("peakRefWavelengthNm", 1100.0),
        center_freq_mhz=_fp("centerFreqMhz", 80.0),
        freq_shift_bandwidth_mhz=_fp("freqShiftBandwidthMhz", 15.0),
        requires_rf_drive=ctx.params.get("requiresRfDrive") is True,
    )


def diffract_aom_op(ray_in: BeamRay, ctx: PhysicsOpContext) -> list[BeamRay]:
    order = order_from_context(ctx)

    freq_mhz = _read_rf_frequency_mhz(ctx)
    v_acoustic = (
        _positive_finite_number(ctx.params.get("acousticVelocityMps"))
        or 4200.0
    )
    n = _positive_finite_number(ctx.params.get("refractiveIndex")) or 2.26
    L = ctx.params.get("crystalLengthMm")
    if not isinstance(L, (int, float)):
        L = vec3_distance(
            ctx.face_in.position_mm_body_local,
            ctx.face_out.position_mm_body_local,
        )
    L = float(L)

    theta_b = bragg_angle_rad(ray_in.wavelength_nm, freq_mhz, v_acoustic, n)
    deflect_rad = order * 2 * theta_b

    dir_out = _deflect_along_rf_side(
        ray_in.direction,
        _read_rf_direction_body_local(ctx),
        deflect_rad,
    )

    detune = bragg_detuning_factor(
        ray_in, ctx, order, theta_b, freq_mhz, v_acoustic, n, L,
    )
    first_order_eff = first_order_efficiency_from_context(ray_in, ctx, theta_b) * detune
    eff = order_efficiency(order, first_order_eff)
    new_power = ray_in.power_mw * eff

    # Doppler shift: order m diffracts off the f_RF acoustic wave, shifting the
    # optical frequency by m*f_RF. Tracked as an offset on the nominal carrier.
    new_freq_offset_hz = ray_in.freq_offset_hz + order * freq_mhz * 1e6

    b_slab = L / n
    qx_out = _apply_abcd_to_q(1, b_slab, 0, 1, ray_in.qx)
    qy_out = _apply_abcd_to_q(1, b_slab, 0, 1, ray_in.qy)

    new_origin = ctx.face_out.position_mm_body_local
    thickness = vec3_distance(
        ctx.face_in.position_mm_body_local,
        ctx.face_out.position_mm_body_local,
    )

    return [ray_in.replaced(
        origin=new_origin,
        direction=dir_out,
        qx=qx_out,
        qy=qy_out,
        power_mw=new_power,
        path_length_mm=ray_in.path_length_mm + thickness,
        freq_offset_hz=new_freq_offset_hz,
    )]


register_kind("aom", KindEntry(
    ops={"diffract_aom": diffract_aom_op},
    needs_aperture=True,
))
